std-bits/table/src/main/java/org/enso/table/read/ExcelReader.java

```diff
@@ -0,0 +1,408 @@
+      InputStream stream,
+      ExcelRange excelRange,
+      HeaderBehavior headers,
+      int skip_rows,
+      Integer row_limit,
+      boolean xls_format)
+      throws IOException {
+    return readRange(getWorkbook(stream, xls_format), excelRange, headers, skip_rows, row_limit);
+  }
+
+  /** Specifies how to set the headers for the returned table. */
+  public enum HeaderBehavior {
+    /** Tries to infer if the headers are present in the file. */
+    INFER,
+
+    /** Uses the first row in the file as headers. Duplicate names will be appended suffixes. */
+    USE_FIRST_ROW_AS_HEADERS,
+
+    /** Uses the default Excel Column Names (e.g. A, B, C). */
+    EXCEL_COLUMN_NAMES
+  }
+
+  private static Workbook getWorkbook(InputStream stream, boolean xls_format) throws IOException {
+    return xls_format ? new HSSFWorkbook(stream) : new XSSFWorkbook(stream);
+  }
+
+  private static Table readRange(
+      Workbook workbook, ExcelRange excelRange, HeaderBehavior headers, int skip_rows, Integer row_limit) {
+    int sheetIndex = getSheetIndex(workbook, excelRange.getSheetName());
+    if (sheetIndex == -1) {
+      throw new IllegalArgumentException("Unknown sheet '" + excelRange.getSheetName() + "'.");
+    }
+
+    return readTable(
+        workbook,
+        sheetIndex,
+        excelRange,
+        headers,
+        skip_rows,
+        row_limit == null ? Integer.MAX_VALUE : row_limit);
+  }
+
+  private static Table readTable(
+      Workbook workbook,
+      int sheetIndex,
+      ExcelRange excelRange,
+      HeaderBehavior headers,
+      int skipRows,
+      int rowCount) {
+    ExcelSheet sheet = new ExcelSheet(workbook, sheetIndex);
+
+    // Expand Single Cell
+    if (excelRange != null && excelRange.isSingleCell()) {
+      ExcelRow currentRow = sheet.get(excelRange.getTopRow());
+      if (currentRow == null || currentRow.isEmpty(excelRange.getLeftColumn())) {
+        return new Table(
+            new Column[] {
+                new Column(
+                    CellReference.convertNumToColString(excelRange.getLeftColumn() - 1),
+                    new ObjectStorage(new Object[0], 0)
+                )});
+      }
+
+      excelRange = expandSingleCell(excelRange, sheet, currentRow);
+    }
+
+    // Row Range
+    boolean wholeColumn = excelRange == null || excelRange.isWholeColumn();
+    int startRow = (wholeColumn ? 1 : excelRange.getTopRow()) + skipRows;
+    int endRow = wholeColumn ? sheet.getLastRow() : excelRange.getBottomRow();
+
+    // Column Range
+    boolean wholeRow = excelRange == null || excelRange.isWholeRow();
+    int startCol = wholeRow ? 1 : excelRange.getLeftColumn();
+    int endCol = wholeRow ? -1 : excelRange.getRightColumn();
+
+    // Headers
+    NameDeduplicator deduplicator = new NameDeduplicator();
+    String[] columnNames = getHeaders(headers, sheet, startRow, endRow, startCol, endCol, deduplicator);
+    if (columnNames != null) {
+      startRow++;
+    }
```

Comment:
I don't like the fact that we have an implicit invariant that returning `null` determines whether we have consumed the first row or not - these are correlated but they are independent concepts.

Maybe instead we can make the return value richer and indicate whether the first row was consumed by some `rowOffset` value or something?

Then we could also avoid returning `null` for `columnNames` and instead just already return the excel column names in a more uniform manner as the others. We could then also perform the resolution of missing names somewhere near this operation - currently it is done at the end and it is easy to miss that we are doing it at all. And also the `getColumnName` is slightly misleading to me - it sounds like a pure function that I can call as much as I want, but it is actually stateful because it will give a different result on each call with an empty/null argument. Thus I'd suggest ideally to move missing name resolution to a separate preprocess step just right after `getHeaders`. If not, I'd at least suggest renaming `getColumnName` into something like `generateColumnName` to show more clearly that it is in fact a non-pure stateful function.